client/client_pool/include/client/client_pool/client_pool_timer.hpp

```diff
@@ -48,6 +70,16 @@ class Timer {
     timer_thread_future_.wait();
   }
 
+  /*
+   * Start the timer for a timeout, and make an asynchronous call to the on_timeout function (given at this Timer's
+   * construction) with client as a parameter after the timeout duration (also given at this Timer's construction).
+   * Note that, though this Timer's implementation should make a reasonable effort to make this callback after about
+   * the given timeout duration, no hard guarantees about the precise timing of this callback are provided. No callback
+   * will be made if this Timer was constructed with a timeout of 0 milliseconds or if stopTimerThread has previously
+   * been called for this Timer. Behavior is undefined if start is called at any time when there is an ongoing timeout
+   * from a previous call to start that has not either completed its callback or been successfully cancelled via
```

Comment:
I would also be interested to hear @yuliasherman's take on this if she is familiar with the intended API here, as the comments I have added to `client_pool_timer.hpp` and the re-worked tests include some interpretations I am not completely certain about.

Reply:
I will say that, under my current unconfirmed interpretations, I think the API does allow for successive `start` calls if the callbacks from all previous `start` calls have either already been completed or have been _successfully_ cancelled with `cancel`. The case where I have interpreted behavior as undefined is specifically the one where there is a callback scheduled by a previous `start` call still in flight at the time a new `start` call begins. This includes both the case where `start` is called once again in a row without cancelling or waiting for the previous callback, and the case where the previous call was cancelled with `cancel`, but that cancellation failed.

I will mention I think it is non-ideal that, at least under my current interpretation, `Timer::cancel` can effectively fail silently and thereby leave the `Timer` in a state where subsequent calls to `start` have undefined behavior. I am not quite sure what the best way to rectify this would be, though I might be interested to examine that further and try to produce a follow-up PR to clean this up.

Reply:
@upshaw-alex I think the explanation is good enough. There is no way "to make sure that no successive start calls for the same client are made".

Reply:
Thanks for clearing that up @yuliasherman!